### PR TITLE
requirements file and git ignore added

### LINKED ISSUE #1: Refactor your notebook to use functions imported from modules
### Problem to solve

As `scientist`, I want `to re-se functions` so that `my notebooks are better organized and more maintainable`.

### Proposal

- [x] Refactor re-usable code into functions/classes, which includes
  - [x] Plots used for analysis
    - [x] Histograms (test/train set)
    - [x] ROC curve
    - [x] confusion matrix
    - [x] Summary of matrics 
  - [x] Training a transformer classifer
  - [x] ....
- [x] Add them to modules with meaningful names in your repository
- [x] add `requirement.txt` to run all modules

### What does success look like, and how can we measure that?

- [ ] Functions are implemented and used in the notebook `06TestModel.ipynb`

- weight: 3

### PR DESCRIPTION
**Pull request by rajhaq**

### CODE REVIEW

File: scripts/plotting/__pycache__/confusion_matrix.cpython-310.pyc
Comment (kristian4cast) [inline scripts/plotting/__pycache__/confusion_matrix.cpython-310.pyc:1]: remove whole folder __pycache__ and add to .gitignore

File: 01deBERTaModel.ipynb [hunk 1]
```diff
@@ -0,0 +1,490 @@
+{
+ "cells": [
```
Comment (kristian4cast) [inline 01deBERTaModel.ipynb:2]: add headline for whole notebook. what is being done in the notebook, what are the results,....

File: scripts/plotting/confusion_matrix.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    - figsize: tuple specifying the figure size
+    """
+    cm = confusion_matrix(y_true, y_pred)
+    plt.figure(figsize=figsize)
+    
+    sns.heatmap(cm, annot=annot, fmt=fmt, cmap=cmap, cbar=False, square=True, linewidths=0.5, linecolor='black', xticklabels=labels, yticklabels=labels)
+    
+    plt.title(title)
+    plt.xlabel('Predicted Label')
+    plt.ylabel('True Label')
```
Comment (kristian4cast) [inline scripts/plotting/confusion_matrix.py]: maybe introduce function that takes care of these steps as they are required by basically every plotting function in your library

File: scripts/plotting/confusion_matrix.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    plt.show()
```
Comment (kristian4cast) [inline scripts/plotting/confusion_matrix.py]: shouldn't be part of the plotting function, may cause annoying behavior later....

File: scripts/plotting/dataset_length_distribution.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import seaborn as sns
+import matplotlib.pyplot as plt
+
+def plot_numeric_distribution(df, column, bins=30, title='Distribution of Numeric Feature', x_label='Feature Value', y_label='Frequency'):
+    sns.histplot(df[column].apply(len), bins=bins)
+    plt.title(title)
+    plt.xlabel(x_label)
+    plt.ylabel(y_label)
```
Comment (kristian4cast) [inline scripts/plotting/dataset_length_distribution.py]: use function

File: scripts/plotting/test_training_distribution.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import matplotlib.pyplot as plt
+import xarray as xr
+
+def plot_label_distribution_split(data, split_indices, column='Label', bins=20, colors=['blue', 'green'], alpha=0.7, titles=['Training Set Label Distribution', 'Test Set Label Distribution'], x_label='Label', y_label='Frequency', figsize=(12, 6)):
```
Comment (kristian4cast) [inline scripts/plotting/test_training_distribution.py]: probably a code formatter should be introduced like [black](https://github.com/psf/black). In my project and at 4cast in general we use [pre-commit](https://pre-commit.com/) to handle these things for us. You could basically use my [pre-commit config file](https://gitlab.com/4cast1/maelstrom/a2/a2/-/blob/main/.pre-commit-config.yaml?ref_type=heads) and start from there. 


Basically you need
1. to install it  `pip install pre-commit`
2. put the "pre-commit config file" .pre-commit-config.yaml in your project's base folder
3. Install it `pre-commit install`
4. Run it on all code `pre-commit run --all`
5. From now all new commits will automatically be checked by `pre-commit`

add `pyproject.toml` file:
```yaml
[tool.flake8]
max-line-length = 120
per-file-ignores = [
    "__init__.py:F401,E501",
    "training_deep500.py:F403,F405",
    "training_performance.py:F403,F405",
    "scripts/relevance_dataset_generation/mlflow_projects/generate_relevance_labels.py:E501",
]

[tool.black]
line-length = 120


[tool.vulture]
min_confidence = 80
paths = ['src/']
exclude = [
    '**/cachedir/*',
    '**/.venv/*',
    'src/a2/training/tracking_hugging.py',
    'src/a2/training/training_performance.py',
    'src/a2/training/training_deep500.py',
    'src/tests/unit/twitter/test_locations.py',
]
ignore_names = ["control", "gpu"]
sort_by_size = true
verbose = false

Comment (kristian4cast) [inline scripts/plotting/test_training_distribution.py]: maybe give `data` a more meaningful name and give it a type. also `split_indices` should have a type

File: scripts/transformer_trainer.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+def compute_metrics(eval_pred):
+    predictions, labels = eval_pred
+    predictions = predictions.argmax(axis=-1)
+    classification_report = sklearn.metrics.classification_report(
+        labels, predictions, target_names=["not raining", "raining"], output_dict=True
+    )
+    f1_not_raining = classification_report["not raining"]["f1-score"]
+    f1_raining = classification_report["raining"]["f1-score"]
+    return {"f1_not_raining": f1_not_raining, "f1_raining": f1_raining}
+def get_trainer(dataset, db_config_base, model_nm, FOLDER_TO_OUTPUT, parameters, tokenizer):
```
Comment (kristian4cast) [inline scripts/transformer_trainer.py]:
```suggestion
def get_trainer(dataset, db_config_base, model_nm, folder_to_output, parameters, tokenizer):
```
variable names should be lower case

File: scripts/transformer_trainer.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    return transformers.Trainer(
+        model_init=get_model_partial,
+        args=args,
+        train_dataset=dataset["train"],
+        eval_dataset=dataset["test"],
+        tokenizer=tokenizer,
+        compute_metrics=compute_metrics,
+    )
+
+def run_training(parameters, model_nm, dataset, FOLDER_TO_OUTPUT, tokenizer):
```
Comment (kristian4cast) [inline scripts/transformer_trainer.py]:
```suggestion
def run_training(parameters, model_nm, dataset, folder_to_output, tokenizer):
```
same as above

**Review by kristian4cast — CHANGES_REQUESTED** [review]

File: src/plotting/plot.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+    :param title: The title of the plot.
+    :param legend_loc: The location of the legend on the plot. If None, the legend is not shown.
+    """
+    plt.xlabel(xlabel)
+    plt.ylabel(ylabel)
+    plt.title(title)
+    if legend_loc is not None:
+        plt.legend(loc=legend_loc)
+    plt.show()  # This would display the plot
```
Comment (kristian4cast) [inline src/plotting/plot.py]: remove all `plt.show()` from your library. i would just use these in notebooks. otherwise they cause issues

File: src/plotting/plot.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import matplotlib.pyplot as plt
+
+
+def plot(title, xlabel="X", ylabel="Y", legend_loc=None):
```
Comment (kristian4cast) [inline src/plotting/plot.py]: rename to something more meaningful, e.g. `adjust_plot`

**Review by kristian4cast — CHANGES_REQUESTED** [review]

**Review by kristian4cast — APPROVED** [review]
